language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/KeySpecifierNodeContext.java

```diff
@@ -89,23 +92,49 @@ public KeySpecifierNodeContext() {
+            }
         }
-        if (rowTypeSymbol.typeKind() != TypeDescKind.RECORD) {
-            return completionItems;
-        }
-        RecordTypeSymbol recordTypeSymbol = (RecordTypeSymbol) rowTypeSymbol;
+
         // Get existing keys
         Set<String> fieldNames = node.fieldNames().stream()
                 .filter(identifierToken -> !identifierToken.isMissing())
                 .map(Token::text)
                 .collect(Collectors.toSet());
-        // Get field symbols which are not already specified
-        List<RecordFieldSymbol> symbols = recordTypeSymbol.fieldDescriptors().values().stream()
-                .filter(recordFieldSymbol -> recordFieldSymbol.getName().isPresent()
-                        && SyntaxInfo.isIdentifier(recordFieldSymbol.getName().get())
-                        && !fieldNames.contains(recordFieldSymbol.getName().get()))
-                .collect(Collectors.toList());
-        completionItems.addAll(this.getCompletionItemList(symbols, context));
+        
+        if (rowTypeSymbol.typeKind() == TypeDescKind.RECORD) {
+            RecordTypeSymbol recordTypeSymbol = (RecordTypeSymbol) rowTypeSymbol;
+            // Get field symbols which are not already specified
+            List<RecordFieldSymbol> symbols = recordTypeSymbol.fieldDescriptors().values().stream()
+                    .filter(recordFieldSymbol -> recordFieldSymbol.getName().isPresent()
+                            && SyntaxInfo.isIdentifier(recordFieldSymbol.getName().get())
+                            && !fieldNames.contains(recordFieldSymbol.getName().get()))
+                    .collect(Collectors.toList());
+            completionItems.addAll(this.getCompletionItemList(symbols, context));
+        } else if (rowTypeSymbol.typeKind() == TypeDescKind.UNION && 
+                CommonUtil.isUnionOfType((UnionTypeSymbol) rowTypeSymbol, TypeDescKind.RECORD)) {
+            // If row type is a union of records, we find the common named fields (we don't consider the type of those
+            // fields as of now) and show them in completions.
+            UnionTypeSymbol unionTypeSymbol = (UnionTypeSymbol) rowTypeSymbol;
+            List<RecordFieldSymbol> commonFields = unionTypeSymbol.memberTypeDescriptors().stream()
+                    .map(CommonUtil::getRawType)
+                    .map(typeSymbol -> (RecordTypeSymbol) typeSymbol)
+                    .map(RecordTypeSymbol::fieldDescriptors)
+                    .reduce((map1, map2) -> map1.entrySet().stream()
+                            .filter(e -> map2.containsKey(e.getKey()))
+                            .filter(e -> !fieldNames.contains(e.getKey()))
+                            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue)))
+                    .map(fieldMap -> new ArrayList<>(fieldMap.values()))
+                    .orElse(new ArrayList<>());
+            completionItems.addAll(this.getCompletionItemList(commonFields, context));
+        }
```

Comment:
Shall we move this to a separate utility method for a cleaner approach?

Reply:
Done!